core/engine/src/object/builtins/lazy_builtin.rs

```diff
@@ -0,0 +1,453 @@
+    Context, JsData, JsNativeError, JsObject, JsResult, JsValue, NativeFunction,
+};
+use boa_gc::{Finalize, Trace, WeakGc};
+
+#[derive(Debug, Clone, Trace, Finalize)]
+pub(crate) enum BuiltinKind {
+    Function(NativeFunctionObject),
+    Ordinary,
+}
+
+/// A builtin function. Used for lazy initialization of builtins.
+
+#[derive(Clone, Finalize, Debug)]
+#[allow(clippy::type_complexity)]
+pub struct LazyBuiltIn {
+    pub(crate) init_and_realm: Option<(fn(&Realm), WeakGc<RealmInner>)>,
+    pub(crate) kind: BuiltinKind,
+}
+
+impl LazyBuiltIn {
+    pub(crate) fn set_constructor(&mut self, function: NativeFunction, realm: Realm) {
+        if let BuiltinKind::Function(ref mut native_function) = self.kind {
+            native_function.f = function;
+            native_function.constructor = Some(ConstructorKind::Base);
+            native_function.realm = Some(realm);
+        } else {
+            panic!("Expected BuiltinKind::Function");
+        }
+    }
+}
+
+// SAFETY: Temporary, TODO move back to derived Trace when possible
+unsafe impl Trace for LazyBuiltIn {
+    custom_trace!(this, mark, {
+        mark(&this.kind);
+    });
+}
+
+// Implement the trait for JsData by overriding all internal_methods by calling init before calling into the underlying internel_method
+impl JsData for LazyBuiltIn {
+    fn internal_methods(&self) -> &'static InternalObjectMethods {
+        static FUNCTION: InternalObjectMethods = InternalObjectMethods {
+            __construct__: lazy_construct,
+            __call__: lazy_call,
+            ..LAZY_INTERNAL_METHODS
+        };
+
+        if let BuiltinKind::Function(_) = self.kind {
+            return &FUNCTION;
+        }
+
+        &LAZY_INTERNAL_METHODS
+    }
+}
+
+pub(crate) static LAZY_INTERNAL_METHODS: InternalObjectMethods = InternalObjectMethods {
+    __get_prototype_of__: lazy_get_prototype_of,
+    __set_prototype_of__: lazy_set_prototype_of,
+    __is_extensible__: lazy_is_extensible,
+    __prevent_extensions__: lazy_prevent_extensions,
+    __get_own_property__: lazy_get_own_property,
+    __define_own_property__: lazy_define_own_property,
+    __has_property__: lazy_has_property,
+    __try_get__: lazy_try_get,
+    __get__: lazy_get,
+    __set__: lazy_set,
+    __delete__: lazy_delete,
+    __own_property_keys__: lazy_own_property_keys,
+    __call__: non_existant_call,
+    __construct__: non_existant_construct,
+};
+
+pub(crate) fn lazy_get_prototype_of(
+    obj: &JsObject,
+    context: &mut Context,
+) -> JsResult<JsPrototype> {
+    let builtin: JsObject<LazyBuiltIn> = obj.clone().downcast().expect("obj is not a Builtin");
+    // If there is a value in here, we need to initialize the builtin
+    if builtin.borrow().data.init_and_realm.is_some() {
+        let mut builtin_borrow = builtin.borrow_mut();
+        let (init, realm_inner) = builtin_borrow
+            .data
+            .init_and_realm
+            .take()
+            .expect("init_and_realm not set");
```

Comment:
With the new changes this can be improved. Instead of calling `init_and_realm.is_some()`, then `init_and_realm.take()`, we can do `if let Some(data) = builtin.borrow_mut().data.init_and_realm.take()`